### ISSUE TITLE
www/caddy: Allow writing *.global and *.conf files in the UI

### ISSUE THREAD

**Issue by DaelonSuzuka** [opnsense/plugins#4095]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/plugins/blob/master/CONTRIBUTING.md
- [x] I have searched the existing issues, open and closed, and I'm convinced that mine is new.
- [x] When the request is meant for an existing plugin, I've added its name to the title.

**Is your feature request related to a problem? Please describe.**
I have a caddy LXC in my homelab that I'd like to move into my new opnsense router. Using the UI is *extremely* tedious compared to writing my simple reverse_proxy directives by hand. The entire value proposition of Caddy is that adding a new entry can be 50 characters of plaintext, but the UI requires a TON of duplication and creating matching entries across multiple tabs. 

**Describe the solution you'd like**
It should be straightforward to allow creating and editing `caddy.d/*.global` and `caddy.d/*.conf` files in the UI. It would probably be fine to not even manage multiple files, just one of each with its own textedit widget.

The plugin's Caddyfile is already configured for importing them, and it already has a mechanism to validate the configuration before reloading (not to mention that Caddy is very resilient to bad configs). I don't know opnsense's UI framework but a couple of textedits that sync to disk *should* be very straightforward.

**Comment by Monviech** [comment]
Can't do that, sorry.

There is another plugin version that is in the community maintained repo, where you can just copy paste the whole Caddyfile into the GUI: https://github.com/mimugmail/opn-repo

**Comment by DaelonSuzuka** [comment]
You can't... make the plugin work correctly? I followed all the steps in the docs and it won't retrieve certs for domains/handlers set up through the UI, but the exact same domains work exactly as expected when I put them in a `caddy.d/*.conf` file.

I guess I'll try mimugmail's version.

**Comment by Monviech** [comment]
From what I know the plugin works as intended. Do you have any debug logs that show the certificate requests failing when domains have been set up through the UI?

Please share them with me so I can see what the problem is.

**Comment by DaelonSuzuka** [comment]
> From what I know the plugin works as intended. Do you have any debug logs that show the certificate requests failing when domains have been set up through the UI?

It wasn't the cert request failing, I think it might have been a caching problem with reusing already registered certs? It's working now, but only after several cycles of un/reinstalling the plugin, and manually deleting the certs folder to force it to re-request them all. At this point it was probably user error, but I followed the docs multiple times over a couple days this weekend and I don't see a reason why it wasn't working but is working now.

My only complaint now is that it's rather tedious to flip back and forth between the `Domains` and `Handlers` tabs since I only actually care about one field on each side. The mandatory description field is also annoying, since I don't see a reason that an empty description couldn't just automatically take the domain name or something. Between these two things it takes like 5x as much interaction to create a reverse proxy endpoint as it does to type `<domain> { reverse_proxy <target> }` into a caddyfile. 

Maybe a "create matching handler" button in the "commands" section of the domains/subdomains table?

I did try mimugmail's `os-caddy-maxit`, but it looks like `os-caddy` doesn't uninstall properly (purely guessing that something is left behind, maybe the UI tabs are defined by directories, and the package uninstaller leaves empty directories?):

<details>
<summary>screenshot</summary>

![vivaldi_CUc2IdVgf6](https://github.com/user-attachments/assets/55ddc9a0-3ff0-4652-b5e7-2cc7390073e4)

</details>

> [docs.opnsense.org/development/guidelines/basics.html](https://docs.opnsense.org/development/guidelines/basics.html#safeguard-user-input)
>
> If you want to make a proper argument please respect the boundaries of the project.

Maybe that link should be in `CONTRIBUTING.md` or the issue template.

**Comment by Monviech** [comment]
I couldn't make the GUI simpler since it has a lot of options.

One of my first plugin versions only had one page where you could create only domain/reverse_proxy entries without any other options. https://github.com/Monviech/os-caddy-plugin/tree/1.0.8b

But since the options inside the forms are quite plentiful now, that was not possible anymore.

I'm looking into your feedback, maybe I can improve something, but I don't know yet. Most Caddyfiles of users I have seen so far were pretty simple and static.

**Comment by DaelonSuzuka** [comment]
Yeah, I think it's a friction and discoverability problem, not a structural issue. I mostly just use `reverse_proxy` directives, so after I created the first Domain object, it wasn't obvious what I'm supposed to do next. I eventually remembered reading in the Caddy docs that `reverse_proxy` is basically a macro that implicitly expands into a block of simpler directives, and then Handlers made sense.

Some way to nudge the user from a Domain to creating a matching Handler (prefilling the Domain we came from) ([0]) would:
- reduce the amount of motion, making the common path faster for power users without actually removing any options
- limit the cognitive load for the new user by making it more obvious what's "supposed" to be next

<details>
<summary>[0]</summary>

and not distracting them with a mandatory irrelevant description field that I don't care about when I just want to reverse proxy something and I'm trying to figure out if it doesn't work because of my DNS or my firewall rules ~~or my inability to read~~

</details>

**Comment by DaelonSuzuka** [comment]
> If you have a good idea here, feel free to propose something in a PR that makes it easier, I'd be interested.

I should have some time this week; I'll see what I can put together!

> \< info >

Thanks for the links, that looks like a good starting point.

**Comment by Monviech** [comment]
Thanks. If you need any more info feel free to ask when you get stuck.

**Comment by Monviech** [comment]
@DaelonSuzuka FYI https://github.com/opnsense/plugins/pull/4097

**Comment by Monviech** [comment]
@DaelonSuzuka What do you think of something really simple like this? It would reduce friction and save a few button clicks, and there is no extensive logic inside of it that would break in the future, like pre filling forms etc... its really easy to understand for the user and for the dev.

https://github.com/Monviech/opnsense-plugins/commit/e7a360a2f42e383effe621235aa246adb4220024

**Comment by DaelonSuzuka** [comment]
Hey @Monviech, I meant to get back to this way sooner, but my new job started the week after and I got completely busy.

I've installed those updates to the caddy plugin and they make a big difference! The step 1 and step 2 buttons are a great solution, especially given the constraints of the UI framework.

I think it's a LOT more approachable for someone who's mostly written simple Caddyfiles and hasn't used the JSON API or dug into the internals (aka me). Thanks a lot!
